{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import sys\n",
    "sys.path.append('/Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent')\n",
    "from experiments.utilities.evaluation import AgentEvaluator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_set = {\n",
    "    \"SwissLipids\": [\"5.ttl\"],\n",
    "    \"Uniprot\": [\"51_all_proteins_linked_to_arachidonate.ttl\", \"50_tissues_where_genes_sphingosines_are_expressed.ttl\"],\n",
    "    \"Rhea\": [\"12.ttl\", \"112_Total_number_of_reaction_ec_protein_links.ttl\", \"117_Retrieve_parent_reactions_and_their_catalyzing_enzymes_for_a_metabolite_that_is_not_directly_involved_in_a_reaction.ttl\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing endpoint set: SwissLipids\n",
      "\n",
      "Processed: 5.ttl\n",
      "Processed 1 files from SwissLipids\n",
      "Processing endpoint set: Uniprot\n",
      "\n",
      "Processed: 50_tissues_where_genes_sphingosines_are_expressed.ttl\n",
      "\n",
      "Processed: 51_all_proteins_linked_to_arachidonate.ttl\n",
      "Processed 2 files from Uniprot\n",
      "Processing endpoint set: Rhea\n",
      "\n",
      "Processed: 12.ttl\n",
      "\n",
      "Processed: 112_Total_number_of_reaction_ec_protein_links.ttl\n",
      "\n",
      "Processed: 117_Retrieve_parent_reactions_and_their_catalyzing_enzymes_for_a_metabolite_that_is_not_directly_involved_in_a_reaction.ttl\n",
      "Processed 3 files from Rhea\n",
      "Saved metadata for 6 files from 3 endpoint sets to:\n",
      "  - /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/testset_meta_data.json\n",
      "Sending question 1/6 to agent...\n",
      "{'messages': [HumanMessage(content='For a given lipid (SWISSLIPID:000399814, Ceramide), return a list of all corresponding reactions (Rhea IDs and equation).\\nThis query does consider the hierarchy: it looks for reactions involving the specified lipid and/or its more specific forms.', additional_kwargs={}, response_metadata={}, id='cecf4cfb-b766-4ee1-811b-780ab08ab541')], 'structured_question': StructuredQuestion(intent='access_resources', extracted_classes=['lipid', 'reaction'], extracted_entities=['SWISSLIPID:000399814', 'Ceramide', 'Rhea IDs', 'equation'], question_steps=['Get the lipid entity for SWISSLIPID:000399814', 'Get all subclass lipids of the lipid entity', 'Get all reactions involving the lipid entity and/or its subclasses', 'Return Rhea IDs and equations for the reactions']), 'retrieved_docs': [], 'extracted_entities': [{'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_61678', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7403080463409424, 'endpoint_url': ''}, page_content='(+)-endo-beta-bergamotene'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_229209', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.739495575428009, 'endpoint_url': ''}, page_content='FR-900098(1-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_78904', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7387851476669312, 'endpoint_url': ''}, page_content='4,4-dimethyl-8,14-cholestadien-3beta-ol'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_86131', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7384021282196045, 'endpoint_url': ''}, page_content='5alpha-cholesta-8,14-dien-3beta-ol'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_49204', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7378146052360535, 'endpoint_url': ''}, page_content='(+)-guaia-6,9-diene')], 'text': 'SWISSLIPID:000399814'}, {'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_17761', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 1.0, 'endpoint_url': ''}, page_content='ceramide'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_71986', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8857792615890503, 'endpoint_url': ''}, page_content='C20 ceramide'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_84404', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8758777379989624, 'endpoint_url': ''}, page_content='ceramide 1-phosphate(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_74997', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8397696018218994, 'endpoint_url': ''}, page_content='mannosyl diphosphorylinositol ceramide(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_73202', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8332492113113403, 'endpoint_url': ''}, page_content='ceramide phosphoethanolamine zwitterion')], 'text': 'Ceramide'}, {'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_138433', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.71418297290802, 'endpoint_url': ''}, page_content='(9S,13S)-1a,1b-dinor-12-oxo-10,15-phytodienoyl-CoA(4-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_138431', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7066922783851624, 'endpoint_url': ''}, page_content='(9S,13S)-1a,1b-dinor-10,11-dihydro-12-oxo-15-phytoenoyl-CoA(4-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_62338', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.703957200050354, 'endpoint_url': ''}, page_content='O(4)-phosphonato-L-tyrosine(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_58354', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7029036283493042, 'endpoint_url': ''}, page_content='4-amino-2-methyl-5-phosphonatooxymethylpyrimidine(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_57890', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6983885765075684, 'endpoint_url': ''}, page_content=\"adenosine 5'-phosphoramidate(1-)\")], 'text': 'Rhea IDs'}, {'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_90781', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7454793453216553, 'endpoint_url': ''}, page_content='13-oxo-9Z,11E-ODE(1-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_194452', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7307087182998657, 'endpoint_url': ''}, page_content='5-oxo-ODE(1-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_143142', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6928855180740356, 'endpoint_url': ''}, page_content=\"N,N',N'',N'''-tetraacetylchitotetraose\"), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_190693', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6915227174758911, 'endpoint_url': ''}, page_content='substance P(1-9)(2+)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_131975', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6913023591041565, 'endpoint_url': ''}, page_content='(8R,9S)-EET(1-)')], 'text': 'equation'}], 'structured_output': {'query': '# https://query.wikidata.org/sparql\\nPREFIX wdt: <http://www.wikidata.org/prop/direct/>\\nPREFIX wd: <http://www.wikidata.org/entity/>\\n\\nSELECT ?lipidEntity ?lipidEntityLabel \\nWHERE {\\n  SERVICE <https://query.wikidata.org/sparql> {\\n    ?lipidEntity wdt:P6612 \"SWISSLIPID:000399814\" .\\n    ?lipidEntity rdfs:label ?lipidEntityLabel .\\n    FILTER(LANG(?lipidEntityLabel) = \"en\") .\\n  }\\n}', 'endpoint_url': 'https://query.wikidata.org/sparql'}, 'steps': [StepOutput(label='Generated SPARQL query', details='\\t\\t\\t\\t\\t\\t - Make sure to adhere to the schema and structure of the data as implied by the context (e.g., specific ontologies or data models mentioned).\\n\\n**User Question:**\\nGet the lipid entity for SWISSLIPID:000399814\\n\\n**Classes:**\\n[\\'Lipid\\', \\'Entity\\']\\n\\n**Entities:**\\n[\\'SWISSLIPID:000399814\\']\\n\\n**Endpoints:**\\n{\\'swisslipid\\': \\'https://query.wikidata.org/sparql\\'}\\n\\n**Potential Entities Extracted from the User Question:**\\n[{\\'matchs\\': [Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_61678\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7403080463409424, \\'endpoint_url\\': \\'\\'}, page_content=\\'(+)-endo-beta-bergamotene\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_229209\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.739495575428009, \\'endpoint_url\\': \\'\\'}, page_content=\\'FR-900098(1-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_78904\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7387851476669312, \\'endpoint_url\\': \\'\\'}, page_content=\\'4,4-dimethyl-8,14-cholestadien-3beta-ol\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_86131\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7384021282196045, \\'endpoint_url\\': \\'\\'}, page_content=\\'5alpha-cholesta-8,14-dien-3beta-ol\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_49204\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7378146052360535, \\'endpoint_url\\': \\'\\'}, page_content=\\'(+)-guaia-6,9-diene\\')], \\'text\\': \\'SWISSLIPID:000399814\\'}, {\\'matchs\\': [Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_17761\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 1.0, \\'endpoint_url\\': \\'\\'}, page_content=\\'ceramide\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_71986\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.8857792615890503, \\'endpoint_url\\': \\'\\'}, page_content=\\'C20 ceramide\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_84404\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.8758777379989624, \\'endpoint_url\\': \\'\\'}, page_content=\\'ceramide 1-phosphate(2-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_74997\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.8397696018218994, \\'endpoint_url\\': \\'\\'}, page_content=\\'mannosyl diphosphorylinositol ceramide(2-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_73202\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.8332492113113403, \\'endpoint_url\\': \\'\\'}, page_content=\\'ceramide phosphoethanolamine zwitterion\\')], \\'text\\': \\'Ceramide\\'}, {\\'matchs\\': [Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_138433\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.71418297290802, \\'endpoint_url\\': \\'\\'}, page_content=\\'(9S,13S)-1a,1b-dinor-12-oxo-10,15-phytodienoyl-CoA(4-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_138431\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7066922783851624, \\'endpoint_url\\': \\'\\'}, page_content=\\'(9S,13S)-1a,1b-dinor-10,11-dihydro-12-oxo-15-phytoenoyl-CoA(4-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_62338\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.703957200050354, \\'endpoint_url\\': \\'\\'}, page_content=\\'O(4)-phosphonato-L-tyrosine(2-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_58354\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7029036283493042, \\'endpoint_url\\': \\'\\'}, page_content=\\'4-amino-2-methyl-5-phosphonatooxymethylpyrimidine(2-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_57890\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.6983885765075684, \\'endpoint_url\\': \\'\\'}, page_content=\"adenosine 5\\'-phosphoramidate(1-)\")], \\'text\\': \\'Rhea IDs\\'}, {\\'matchs\\': [Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_90781\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7454793453216553, \\'endpoint_url\\': \\'\\'}, page_content=\\'13-oxo-9Z,11E-ODE(1-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_194452\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.7307087182998657, \\'endpoint_url\\': \\'\\'}, page_content=\\'5-oxo-ODE(1-)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_143142\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.6928855180740356, \\'endpoint_url\\': \\'\\'}, page_content=\"N,N\\',N\\'\\',N\\'\\'\\'-tetraacetylchitotetraose\"), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_190693\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.6915227174758911, \\'endpoint_url\\': \\'\\'}, page_content=\\'substance P(1-9)(2+)\\'), Document(metadata={\\'uri\\': \\'http://purl.obolibrary.org/obo/CHEBI_131975\\', \\'type\\': \\'literal\\', \\'description\\': \\'Chemical entities in ChEBI.\\', \\'score\\': 0.6913023591041565, \\'endpoint_url\\': \\'\\'}, page_content=\\'(8R,9S)-EET(1-)\\')], \\'text\\': \\'equation\\'}]\\n```sparql\\n# https://query.wikidata.org/sparql\\nPREFIX wdt: <http://www.wikidata.org/prop/direct/>\\nPREFIX wd: <http://www.wikidata.org/entity/>\\n\\nSELECT ?lipidEntity ?lipidEntityLabel \\nWHERE {\\n  SERVICE <https://query.wikidata.org/sparql> {\\n    ?lipidEntity wdt:P6612 \"SWISSLIPID:000399814\" .\\n    ?lipidEntity rdfs:label ?lipidEntityLabel .\\n    FILTER(LANG(?lipidEntityLabel) = \"en\") .\\n  }\\n}\\n```', substeps=[], type='context', fixed_message=None)]}\n",
      "Got first run!\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Result: {'final_state_response': '# https://query.wikidata.org/sparql\\nPREFIX wdt: <http://www.wikidata.org/prop/direct/>\\nPREFIX wd: <http://www.wikidata.org/entity/>\\n\\nSELECT ?lipidEntity ?lipidEntityLabel \\nWHERE {\\n  SERVICE <https://query.wikidata.org/sparql> {\\n    ?lipidEntity wdt:P6612 \"SWISSLIPID:000399814\" .\\n    ?lipidEntity rdfs:label ?lipidEntityLabel .\\n    FILTER(LANG(?lipidEntityLabel) = \"en\") .\\n  }\\n}', 'run_id_langsmith': '4e00efcc-2eee-4bbe-8f00-757972a08097', 'in_dataset': False, 'execution_time': '0', 'sparql_construction_prompt_tokens': 1925, 'sparql_construction_completion_tokens': 0, 'sparql_construction_total_tokens': 1925, 'sparql_construction_prompt_cost': 0, 'sparql_construction_completion_cost': 0, 'sparql_construction_total_cost': 0, 'prompt_tokens': 2502, 'completion_tokens': 119, 'total_tokens': 2621, 'prompt_cost': 0, 'completion_cost': 0, 'total_cost': 0}\n",
      "Got result\n",
      "Sending question 2/6 to agent...\n",
      "{'messages': [HumanMessage(content='Where are the human genes encoding enzymes metabolizing N-acyl\\nsphingosines expressed in the human body (federated query, with Rhea and Bgee)', additional_kwargs={}, response_metadata={}, id='41547b3b-da9d-449e-8721-6c605bef1a85')], 'structured_question': StructuredQuestion(intent='access_resources', extracted_classes=['human genes', 'enzymes', 'N-acyl sphingosines'], extracted_entities=['Rhea', 'Bgee'], question_steps=['Find human genes encoding enzymes', 'Identify enzymes metabolizing N-acyl sphingosines', 'Determine expression in the human body', 'Perform federated query with Rhea and Bgee']), 'retrieved_docs': [], 'extracted_entities': [{'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_16708', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.71465003490448, 'endpoint_url': ''}, page_content='adenine'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_16335', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7063755393028259, 'endpoint_url': ''}, page_content='adenosine'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_62338', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7004866003990173, 'endpoint_url': ''}, page_content='O(4)-phosphonato-L-tyrosine(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_145541', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6950992941856384, 'endpoint_url': ''}, page_content=\"adenosine 3'-diphosphate,5'-phosphate(5-)\"), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_57524', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6926901340484619, 'endpoint_url': ''}, page_content='O-phosphonato-L-serine(2-)')], 'text': 'Rhea'}, {'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_79628', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6885541677474976, 'endpoint_url': ''}, page_content='gardenin B'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_4648', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6526753306388855, 'endpoint_url': ''}, page_content='hop-22(29)-ene'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_145921', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.650384247303009, 'endpoint_url': ''}, page_content='bifonsecin B'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_77869', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6469536423683167, 'endpoint_url': ''}, page_content='(E,E)-sorbate'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_133455', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6463495492935181, 'endpoint_url': ''}, page_content='(2E)-dec-2-enal')], 'text': 'Bgee'}], 'structured_output': {'query': \"# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX bgee: <http://bgee.org/>\\n\\nSELECT DISTINCT ?gene ?enzyme ?reaction\\nWHERE {\\n  # Query Rhea for enzymes and their associated reactions\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:catalyzedBy ?enzyme .\\n    ?enzyme a obo:GO_0003824 . # GO term for enzyme activity\\n  }\\n  \\n  # Use Bgee to link to genes\\n  SERVICE <https://bgee.org/sparql> { \\n    ?gene obo:RO_0002204 ?enzyme . # Relation for 'gene product encoded by'\\n    FILTER EXISTS { ?gene obo:NCBITaxon_9606 ?any } . # Simplified filter for human genes\\n  }\\n}\", 'endpoint_url': 'https://sparql.rhea-db.org/sparql'}, 'steps': [StepOutput(label='Generated SPARQL query', details=\"\\t\\t\\t\\t\\t\\t- If specific schema details or endpoint URLs are not directly given, infer or use general knowledge that fits the context. \\n\\t\\t\\t\\t\\t\\t- Avoid using any external schema or endpoint information not covered in the provided context.\\n\\n6. **Federated Query:** Use the **Service** keyword to federate the query across multiple SPARQL endpoints as appropriate.\\n\\n**Input Context:**\\n- The primary task involves querying for human genes that encode enzymes.\\n- Relevant endpoints and resources include those from the Swiss Institute of Bioinformatics (SIB) and other biomedical resources.\\n- Potential entities and classes extracted from the question include references to Rhea and Bgee, which are related to enzyme reactions and gene expression, respectively.\\n\\nGiven the context and requirements, the task is to construct a federated SPARQL query that effectively retrieves the desired information. \\n\\nTo achieve this, we need to identify the relevant SPARQL endpoints for Rhea and Bgee, and then formulate a query that can be federated across these endpoints to find human genes encoding enzymes.\\n\\nLet's start by identifying the relevant endpoints:\\n1. Rhea: Rhea is a comprehensive database of biochemical reactions. It is likely that the Rhea SPARQL endpoint contains information about enzymes and their associated reactions.\\n2. Bgee: Bgee is a database that provides information on gene expression. It can be used to identify genes that are expressed in humans.\\n\\nTo find human genes encoding enzymes, we can use the Rhea endpoint to identify enzymes and then link this information to genes in the Bgee endpoint or another relevant endpoint that contains gene information.\\n\\nHere's how we can construct the query:\\n\\n```sparql\\n# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX enzyme: <http://purl.obolibrary.org/obo/GO_0003824>\\nPREFIX taxon: <http://purl.obolibrary.org/obo/NCBITaxon_>\\nPREFIX gene: <http://purl.obolibrary.org/obo/SO_0000704>\\n\\nSELECT DISTINCT ?gene ?enzyme ?reaction\\nWHERE {\\n  # Query Rhea for enzymes and their associated reactions\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:catalyzedBy ?enzyme .\\n    ?enzyme a enzyme:enzyme .\\n  }\\n  \\n  # Link enzymes to genes using a relevant endpoint (e.g., Bgee or another gene-centric endpoint)\\n  # For this example, let's assume we're using a hypothetical endpoint that links enzymes to genes\\n  SERVICE <https://example-gene-endpoint.unil.ch/sparql> { \\n    ?gene a gene:gene .\\n    ?gene <http://example.org/encodes> ?enzyme .\\n    ?gene <http://example.org/taxon> taxon:9606 . # Filter for human genes (NCBITaxon:9606)\\n  }\\n}\\n```\\n\\nThis query first identifies enzymes and their associated reactions in the Rhea database. Then, it links these enzymes to genes in a hypothetical gene-centric endpoint, filtering for human genes.\\n\\nLet's refine this query to fit the actual available endpoints and schema.\\n\\n```sparql\\n# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\n\\nSELECT DISTINCT ?gene ?enzyme ?reaction\\nWHERE {\\n  # Query Rhea for enzymes and their associated reactions\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:catalyzedBy ?enzyme .\\n    ?enzyme a obo:GO_0003824 . # GO term for enzyme activity\\n  }\\n  \\n  # Use Bgee to link to genes, assuming Bgee has a SPARQL endpoint that can be queried for gene information\\n  SERVICE <https://bgee.org/sparql> { \\n    ?gene obo:RO_0002204 ?enzyme . # Relation for 'gene product encoded by'\\n    ?gene obo:NCBITaxon_9606 ?taxon . # Filter for human genes\\n  }\\n}\\n```\\n\\nThis refined query uses the Rhea SPARQL endpoint to identify enzymes and their reactions, and then links these enzymes to genes using the Bgee SPARQL endpoint, filtering for human genes.\\n\\nLet's finalize the query based on the given context and known endpoint information.\\n\\n```sparql\\n# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX bgee: <http://bgee.org/>\\n\\nSELECT DISTINCT ?gene ?enzyme ?reaction\\nWHERE {\\n  # Query Rhea for enzymes and their associated reactions\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:catalyzedBy ?enzyme .\\n    ?enzyme a obo:GO_0003824 . # GO term for enzyme activity\\n  }\\n  \\n  # Use Bgee to link to genes\\n  SERVICE <https://bgee.org/sparql> { \\n    ?gene obo:RO_0002204 ?enzyme . # Relation for 'gene product encoded by'\\n    FILTER EXISTS { ?gene obo:NCBITaxon_9606 ?any } . # Simplified filter for human genes\\n  }\\n}\\n```\", substeps=[], type='context', fixed_message=None)]}\n",
      "Got first run!\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Result: {'final_state_response': \"# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX bgee: <http://bgee.org/>\\n\\nSELECT DISTINCT ?gene ?enzyme ?reaction\\nWHERE {\\n  # Query Rhea for enzymes and their associated reactions\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:catalyzedBy ?enzyme .\\n    ?enzyme a obo:GO_0003824 . # GO term for enzyme activity\\n  }\\n  \\n  # Use Bgee to link to genes\\n  SERVICE <https://bgee.org/sparql> { \\n    ?gene obo:RO_0002204 ?enzyme . # Relation for 'gene product encoded by'\\n    FILTER EXISTS { ?gene obo:NCBITaxon_9606 ?any } . # Simplified filter for human genes\\n  }\\n}\", 'run_id_langsmith': '44b35518-4bf0-4403-abea-6b64c04dcd55', 'in_dataset': False, 'execution_time': '0', 'sparql_construction_prompt_tokens': 1034, 'sparql_construction_completion_tokens': 0, 'sparql_construction_total_tokens': 1034, 'sparql_construction_prompt_cost': 0, 'sparql_construction_completion_cost': 0, 'sparql_construction_total_cost': 0, 'prompt_tokens': 1591, 'completion_tokens': 101, 'total_tokens': 1692, 'prompt_cost': 0, 'completion_cost': 0, 'total_cost': 0}\n",
      "Got result\n",
      "Sending question 3/6 to agent...\n",
      "{'messages': [HumanMessage(content='Find all proteins linked to arachidonate (CHEBI:32395)', additional_kwargs={}, response_metadata={}, id='12f2e233-4aeb-4bb4-817c-cba8a432e8b7')], 'structured_question': StructuredQuestion(intent='access_resources', extracted_classes=['protein'], extracted_entities=['arachidonate', 'CHEBI:32395'], question_steps=['Find proteins associated with arachidonate (CHEBI:32395)']), 'retrieved_docs': [], 'extracted_entities': [{'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_32395', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 1.0, 'endpoint_url': ''}, page_content='arachidonate'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_82751', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8486751317977905, 'endpoint_url': ''}, page_content='cholesteryl arachidonate'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_78033', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8468923568725586, 'endpoint_url': ''}, page_content='methyl arachidonate'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_59002', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8391536474227905, 'endpoint_url': ''}, page_content='N-arachidonoylglycinate'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_137830', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.8140722513198853, 'endpoint_url': ''}, page_content='arachidonoyl amine')], 'text': 'arachidonate'}, {'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_57457', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7459557056427002, 'endpoint_url': ''}, page_content='(6S)-5-formyltetrahydrofolate(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_23925', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7363247275352478, 'endpoint_url': ''}, page_content='(+)-5-epi-aristolochene'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_195366', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7346182465553284, 'endpoint_url': ''}, page_content='(6R)-10-formyltetrahydrofolate(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_63196', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7342264652252197, 'endpoint_url': ''}, page_content='(+)-caryolan-1-ol'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_57456', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7329903841018677, 'endpoint_url': ''}, page_content='5-formimidoyltetrahydrofolate(2-)')], 'text': 'CHEBI:32395'}], 'structured_output': {'query': '# https://sparql.rhea-db.org/sparql\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX rhea: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT DISTINCT ?protein ?reaction ?ecnumber\\nWHERE {\\n    # Identify the CHEBI ID for arachidonate\\n    VALUES ?chebi {obo:CHEBI_32395}\\n    \\n    # Query Rhea to find reactions associated with arachidonate\\n    SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rhea:side ?side .\\n        ?side rhea:contains ?chebi .\\n        ?reaction rhea:catalyzedBy ?protein .\\n    }\\n    \\n    # Retrieve EC numbers associated with the reactions\\n    ?reaction obo:RO_0000057 ?ecnumber .\\n}', 'endpoint_url': 'https://sparql.rhea-db.org/sparql'}, 'steps': [StepOutput(label='Generated SPARQL query', details='\\t- If an entity is linked to a particular endpoint, prioritize using that endpoint.\\n\\t- The generated query should aim to directly leverage the provided endpoint information to construct a federated query.\\n\\n6. **Query Structure:** \\n\\t- Use the SERVICE keyword to construct a federated query.\\n\\t- Ensure the query is well-structured and follows standard SPARQL syntax.\\n\\n**Generated Query:**\\n```sparql\\n# https://sparql.rhea-db.org/sparql\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX rhea: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT DISTINCT ?protein ?reaction ?ecnumber\\nWHERE {\\n    # Identify the CHEBI ID for arachidonate\\n    VALUES ?chebi {obo:CHEBI_32395}\\n    \\n    # Query Rhea to find reactions associated with arachidonate\\n    SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rhea:side ?side .\\n        ?side rhea:contains ?chebi .\\n        ?reaction rhea:catalyzedBy ?protein .\\n    }\\n    \\n    # Retrieve EC numbers associated with the reactions\\n    ?reaction obo:RO_0000057 ?ecnumber .\\n}\\n```\\n**Explanation:**\\nThe query aims to find proteins associated with arachidonate (CHEBI:32395) by leveraging the Rhea database, which contains information on biochemical reactions. The query first identifies the CHEBI ID for arachidonate and then uses the SERVICE keyword to query the Rhea SPARQL endpoint to find reactions that involve arachidonate. It retrieves the proteins catalyzing these reactions and their associated EC numbers. The results include the protein, the reaction it catalyzes, and the EC number associated with the reaction.', substeps=[], type='context', fixed_message=None)]}\n",
      "Got first run!\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Result: {'final_state_response': '# https://sparql.rhea-db.org/sparql\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX rhea: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT DISTINCT ?protein ?reaction ?ecnumber\\nWHERE {\\n    # Identify the CHEBI ID for arachidonate\\n    VALUES ?chebi {obo:CHEBI_32395}\\n    \\n    # Query Rhea to find reactions associated with arachidonate\\n    SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rhea:side ?side .\\n        ?side rhea:contains ?chebi .\\n        ?reaction rhea:catalyzedBy ?protein .\\n    }\\n    \\n    # Retrieve EC numbers associated with the reactions\\n    ?reaction obo:RO_0000057 ?ecnumber .\\n}', 'run_id_langsmith': '39cc40a1-f59b-4970-bfe4-4e0bfe098235', 'in_dataset': False, 'execution_time': '0', 'sparql_construction_prompt_tokens': 1058, 'sparql_construction_completion_tokens': 0, 'sparql_construction_total_tokens': 1058, 'sparql_construction_prompt_cost': 0, 'sparql_construction_completion_cost': 0, 'sparql_construction_total_cost': 0, 'prompt_tokens': 1598, 'completion_tokens': 64, 'total_tokens': 1662, 'prompt_cost': 0, 'completion_cost': 0, 'total_cost': 0}\n",
      "Got result\n",
      "Sending question 4/6 to agent...\n",
      "{'messages': [HumanMessage(content='Select all Rhea reactions linked to protein sequences (enzymes and transporters) in UniProtKB', additional_kwargs={}, response_metadata={}, id='fc4e312d-a270-44eb-b184-19d96925b86c')], 'structured_question': StructuredQuestion(intent='access_resources', extracted_classes=['Rhea reaction', 'protein sequence', 'enzyme', 'transporter'], extracted_entities=['UniProtKB', 'Rhea'], question_steps=['Select all Rhea reactions', 'Link to protein sequences in UniProtKB', 'enzymes and transporters']), 'retrieved_docs': [], 'extracted_entities': [{'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_61457', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7076636552810669, 'endpoint_url': ''}, page_content='UDP-beta-L-arabinopyranose(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_64351', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6996822357177734, 'endpoint_url': ''}, page_content='beta-D-Manp-(1->4)-D-Glcp'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_31013', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6959225535392761, 'endpoint_url': ''}, page_content='cerotate'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_83836', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6933965682983398, 'endpoint_url': ''}, page_content='UDP-beta-L-rhamnose(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_27906', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6882836818695068, 'endpoint_url': ''}, page_content='ubiquinone-0')], 'text': 'UniProtKB'}, {'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_16708', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.71465003490448, 'endpoint_url': ''}, page_content='adenine'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_16335', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7063755393028259, 'endpoint_url': ''}, page_content='adenosine'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_62338', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7004866003990173, 'endpoint_url': ''}, page_content='O(4)-phosphonato-L-tyrosine(2-)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_145541', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6950992941856384, 'endpoint_url': ''}, page_content=\"adenosine 3'-diphosphate,5'-phosphate(5-)\"), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_57524', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.6926901340484619, 'endpoint_url': ''}, page_content='O-phosphonato-L-serine(2-)')], 'text': 'Rhea'}], 'structured_output': {'query': '# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nSELECT DISTINCT ?reaction \\nWHERE {\\n  ?reaction a rhea:Reaction .\\n}', 'endpoint_url': 'https://sparql.rhea-db.org/sparql'}, 'steps': [StepOutput(label='Generated SPARQL query', details='\\t\\t\\t\\t\\t\\t- Use the provided entity information to construct the query patterns where applicable. \\n\\n6. **SPARQL 1.1 Compliance:** The generated query must be compliant with the SPARQL 1.1 standard, making appropriate use of features like **SERVICE** for federated queries.\\n\\n**Generated Query:**\\n```sparql\\n# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nSELECT DISTINCT ?reaction \\nWHERE {\\n  ?reaction a rhea:Reaction .\\n}\\n```\\n**Explanation:**\\nThe user question asks to \"Select all Rhea reactions.\" To answer this, we need to query the Rhea database, which is accessible via a SPARQL endpoint. The Rhea database uses RDF to represent its data, and reactions are represented as instances of the `rhea:Reaction` class.\\n\\n1. The query starts by defining the prefix `rhea:` for the Rhea namespace (`http://rdf.rhea-db.org/`). This prefix is used to abbreviate the URIs within the query.\\n\\n2. The `SELECT DISTINCT ?reaction` clause is used to retrieve a list of unique Rhea reaction identifiers. The `?reaction` variable represents the reactions we are querying for.\\n\\n3. The `WHERE` clause defines the pattern for the data we are interested in. Here, we specify that `?reaction` should be of type `rhea:Reaction`, effectively filtering the results to include only Rhea reactions.\\n\\n4. Since the Rhea database is accessed through a specific SPARQL endpoint (`https://sparql.rhea-db.org/sparql`), we directly query this endpoint. The query does not require federation to other endpoints as per the user question\\'s requirement.\\n\\nThis query directly answers the user\\'s question by retrieving all Rhea reactions, utilizing the Rhea SPARQL endpoint.', substeps=[], type='context', fixed_message=None)]}\n",
      "Got first run!\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Result: {'final_state_response': '# https://sparql.rhea-db.org/sparql\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nSELECT DISTINCT ?reaction \\nWHERE {\\n  ?reaction a rhea:Reaction .\\n}', 'run_id_langsmith': '14cbd6f5-5764-48e3-a2f4-5aaea0f77fce', 'in_dataset': False, 'execution_time': '0', 'sparql_construction_prompt_tokens': 1045, 'sparql_construction_completion_tokens': 0, 'sparql_construction_total_tokens': 1045, 'sparql_construction_prompt_cost': 0, 'sparql_construction_completion_cost': 0, 'sparql_construction_total_cost': 0, 'prompt_tokens': 1588, 'completion_tokens': 81, 'total_tokens': 1669, 'prompt_cost': 0, 'completion_cost': 0, 'total_cost': 0}\n",
      "Got result\n",
      "Sending question 5/6 to agent...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No entities found in structured question\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'messages': [HumanMessage(content='Total number of <reaction> <ec> <protein> links', additional_kwargs={}, response_metadata={}, id='094c9ddf-191d-49fc-be41-aaab4d25365e')], 'structured_question': StructuredQuestion(intent='access_resources', extracted_classes=['reaction', 'ec', 'protein'], extracted_entities=[], question_steps=['Count the number of links between reaction, ec, and protein']), 'retrieved_docs': [], 'extracted_entities': [], 'structured_output': {'query': '# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (SUM(?count) AS ?total_links)\\nWHERE {\\n  {\\n    SELECT (COUNT(DISTINCT ?reaction) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rh:ec ?ec .\\n        ?reaction rh:status ?status .\\n        FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n      }\\n    }\\n  }\\n  UNION\\n  {\\n    SELECT (COUNT(DISTINCT ?protein) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.uniprot.org/sparql> {\\n        ?protein up:enzyme ?ec .\\n      }\\n    }\\n  }\\n}', 'endpoint_url': 'https://sib-federated.rhea-db.org/sparql'}, 'steps': [StepOutput(label='Generated SPARQL query', details='   - The final query should not contain any prefix other than those defined in the input context (if provided). If no prefix is given, focus on using the full URIs.\\n\\n6. **Relevant Endpoints:** - Identify the most relevant SPARQL endpoints from the context or based on general knowledge about biomedical data sources that could contain the required information.\\n\\n7.**Federated Query Structure:** - Use the **SERVICE** keyword to distribute the query across multiple relevant endpoints.\\n\\n**Input Context:**\\n\\n* **Classes**: reaction, ec, protein\\n* **Endpoints**: \\n    - https://sparql.rhea-db.org/sparql (Rhea)\\n    - https://sparql.uniprot.org/sparql (UniProt)\\n    - https://sib-federated.rhea-db.org/sparql (SIB-Federated)\\n* **Prefix**: \\n    - rr: <http://rdf.rhea-db.org/>\\n    - up: <http://purl.uniprot.org/core/>\\n    - rh: <http://rdf.rhea-db.org/reaction/>\\n\\n**Generated SPARQL Query:**\\n```sparql\\n# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (COUNT(DISTINCT ?link) AS ?num_links)\\nWHERE {\\n  # Query Rhea for reactions and their associated EC numbers\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rh:ec ?ec .\\n    ?reaction rh:status ?status .\\n    FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n    BIND(URI(CONCAT(\"http://example.org/link/reaction_ec/\", STR(?reaction), \"_\", STR(?ec))) AS ?link)\\n  }\\n  \\n  # Query UniProt for proteins associated with EC numbers\\n  SERVICE <https://sparql.uniprot.org/sparql> {\\n    ?protein up:enzyme ?ec .\\n    BIND(URI(CONCAT(\"http://example.org/link/ec_protein/\", STR(?ec), \"_\", STR(?protein))) AS ?link2)\\n  }\\n  \\n  # Combine the links from both queries\\n  VALUES ?link { ?link ?link2 }\\n  \\n  # Count the distinct links\\n}\\n```\\nHowever, the above query doesn\\'t correctly count the number of links between reaction, ec, and protein because it doesn\\'t correctly establish the relationship between the different SERVICE clauses.\\n\\nLet\\'s correct and simplify the query:\\n```sparql\\n# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (COUNT(DISTINCT ?reaction) AS ?reaction_ec_count) (COUNT(DISTINCT ?protein) AS ?ec_protein_count)\\nWHERE {\\n  # Query Rhea for reactions and their associated EC numbers\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rh:ec ?ec .\\n    ?reaction rh:status ?status .\\n    FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n  }\\n  \\n  # Query UniProt for proteins associated with EC numbers\\n  SERVICE <https://sparql.uniprot.org/sparql> {\\n    ?protein up:enzyme ?ec .\\n  }\\n  \\n  # The count of distinct reactions and proteins gives the number of links\\n}\\n```\\nThis query counts the number of distinct reactions associated with EC numbers and the number of distinct proteins associated with EC numbers, effectively giving the count of links between reaction and EC, and between EC and protein.\\n\\nTo directly count the links between reaction, EC, and protein, we need a more complex query that can join these entities directly, which is challenging without a direct relationship between reaction and protein. However, we can infer the total number of links by counting the distinct reactions and proteins that are connected through EC numbers.\\n\\nLet\\'s adjust the query to directly count the links:\\n```sparql\\n# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (COUNT(DISTINCT *) AS ?total_links)\\nWHERE {\\n  # Query Rhea for reactions and their associated EC numbers, and UniProt for proteins associated with these EC numbers\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rh:ec ?ec .\\n    ?reaction rh:status ?status .\\n    FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n    SERVICE <https://sparql.uniprot.org/sparql> {\\n      ?protein up:enzyme ?ec .\\n    }\\n    BIND(CONCAT(STR(?reaction), STR(?protein)) AS ?link)\\n  }\\n  # Count the distinct links\\n}\\n```\\nHowever, nesting SERVICE clauses is not valid in SPARQL 1.1. We need to rethink the query to achieve the desired result.\\n\\nLet\\'s simplify by counting the reaction-EC and EC-protein links separately:\\n```sparql\\n# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT ?reaction_ec_count ?ec_protein_count\\nWHERE {\\n  # Query Rhea for the count of reactions associated with EC numbers\\n  {\\n    SELECT (COUNT(DISTINCT ?reaction) AS ?reaction_ec_count)\\n    WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rh:ec ?ec .\\n        ?reaction rh:status ?status .\\n        FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n      }\\n    }\\n  }\\n  # Query UniProt for the count of proteins associated with EC numbers\\n  {\\n    SELECT (COUNT(DISTINCT ?protein) AS ?ec_protein_count)\\n    WHERE {\\n      SERVICE <https://sparql.uniprot.org/sparql> {\\n        ?protein up:enzyme ?ec .\\n      }\\n    }\\n  }\\n}\\n```\\nThis query provides two counts: the number of distinct reactions linked to EC numbers and the number of distinct proteins linked to EC numbers.\\n\\nTo directly answer the user\\'s question with a single count of links between reaction, EC, and protein, we recognize that the direct links are between reaction-EC and EC-protein. Thus, the total number of links can be considered as the sum of these two types of links.\\n\\nHowever, given the current structure, we can simplify to:\\n```sparql\\n# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (SUM(?count) AS ?total_links)\\nWHERE {\\n  # Query Rhea for the count of reactions associated with EC numbers\\n  {\\n    SELECT (COUNT(DISTINCT ?reaction) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rh:ec ?ec .\\n        ?reaction rh:status ?status .\\n        FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n      }\\n    }\\n  }\\n  UNION\\n  # Query UniProt for the count of proteins associated with EC numbers\\n  {\\n    SELECT (COUNT(DISTINCT ?protein) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.uniprot.org/sparql> {\\n        ?protein up:enzyme ?ec .\\n      }\\n    }\\n  }\\n}\\n```\\nThis query sums the counts of reaction-EC links and EC-protein links, providing a total count of links between the entities. \\n\\nThe final answer is:\\n```sparql\\n# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (SUM(?count) AS ?total_links)\\nWHERE {\\n  {\\n    SELECT (COUNT(DISTINCT ?reaction) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rh:ec ?ec .\\n        ?reaction rh:status ?status .\\n        FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n      }\\n    }\\n  }\\n  UNION\\n  {\\n    SELECT (COUNT(DISTINCT ?protein) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.uniprot.org/sparql> {\\n        ?protein up:enzyme ?ec .\\n      }\\n    }\\n  }\\n}\\n```', substeps=[], type='context', fixed_message=None)]}\n",
      "Got first run!\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Result: {'final_state_response': '# https://sib-federated.rhea-db.org/sparql\\nPREFIX rr: <http://rdf.rhea-db.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/reaction/>\\n\\nSELECT (SUM(?count) AS ?total_links)\\nWHERE {\\n  {\\n    SELECT (COUNT(DISTINCT ?reaction) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rh:ec ?ec .\\n        ?reaction rh:status ?status .\\n        FILTER (?status = \"standard\" || ?status = \"unbalanced\")\\n      }\\n    }\\n  }\\n  UNION\\n  {\\n    SELECT (COUNT(DISTINCT ?protein) AS ?count)\\n    WHERE {\\n      SERVICE <https://sparql.uniprot.org/sparql> {\\n        ?protein up:enzyme ?ec .\\n      }\\n    }\\n  }\\n}', 'run_id_langsmith': 'e84b08c3-cbe3-44d1-a2d0-2a977ab7fc61', 'in_dataset': False, 'execution_time': '0', 'sparql_construction_prompt_tokens': 295, 'sparql_construction_completion_tokens': 0, 'sparql_construction_total_tokens': 295, 'sparql_construction_prompt_cost': 0, 'sparql_construction_completion_cost': 0, 'sparql_construction_total_cost': 0, 'prompt_tokens': 831, 'completion_tokens': 49, 'total_tokens': 880, 'prompt_cost': 0, 'completion_cost': 0, 'total_cost': 0}\n",
      "Got result\n",
      "Sending question 6/6 to agent...\n",
      "{'messages': [HumanMessage(content='Retrieve parent reactions (and their catalyzing enzymes) for a metabolite that is not directly involved in a reaction (ex: CHEBI:83137).', additional_kwargs={}, response_metadata={}, id='110ab557-887c-4d8f-beb9-cb2a3e60a117')], 'structured_question': StructuredQuestion(intent='access_resources', extracted_classes=['Metabolite', 'Reaction', 'Enzyme'], extracted_entities=['CHEBI:83137'], question_steps=['Retrieve parent reactions and their catalyzing enzymes for a given metabolite', 'Check if the metabolite is directly involved in a reaction', 'If not, retrieve parent reactions and their catalyzing enzymes']), 'retrieved_docs': [], 'extracted_entities': [{'matchs': [Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_147415', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7350744009017944, 'endpoint_url': ''}, page_content='[Pyr1]apelin-13(2+)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_83862', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7249436378479004, 'endpoint_url': ''}, page_content='(25R)-3-ketocholest-4-en-26-al'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_231653', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7213397026062012, 'endpoint_url': ''}, page_content='(19E)-cur-19-en-17-al(1+)'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_83863', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7210710644721985, 'endpoint_url': ''}, page_content='(25S)-3-ketocholest-4-en-26-al'), Document(metadata={'uri': 'http://purl.obolibrary.org/obo/CHEBI_63683', 'type': 'literal', 'description': 'Chemical entities in ChEBI.', 'score': 0.7170505523681641, 'endpoint_url': ''}, page_content='(13E)-labda-7,13-dien-15-ol')], 'text': 'CHEBI:83137'}], 'structured_output': {'query': '# https://www.ebi.ac.uk/rdf/services/sparql\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX chebi: <http://purl.obolibrary.org/obo/CHEBI_>\\n\\nSELECT DISTINCT ?parent_reaction ?enzyme ?metabolite\\nWHERE {\\n  # Define the given metabolite\\n  VALUES ?metabolite {obo:CHEBI_83137}\\n  \\n  # Query Rhea to find reactions involving the given metabolite\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:side ?side .\\n    ?side ?p ?metabolite .\\n    ?reaction rhea:equation ?equation .\\n    ?reaction rhea:direction ?direction .\\n    \\n    # Identify the parent reaction\\n    ?parent_reaction rhea:reaction ?reaction .\\n  }\\n  \\n  # Query Rhea to find enzymes catalyzing the parent reaction\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?parent_reaction rhea:catalyzedBy ?enzyme .\\n  }\\n  \\n  # Optionally, fetch additional information about the metabolite from ChEBI\\n  SERVICE <https://www.ebi.ac.uk/rdf/services/sparql> {\\n    ?metabolite obo:IAO_0000115 ?description .\\n  }\\n}', 'endpoint_url': 'https://www.ebi.ac.uk/rdf/services/sparql'}, 'steps': [StepOutput(label='Generated SPARQL query', details=\"   - You may assume there are services/endpoints available that can provide the required data as per the user's question.\\n\\n6.**ChEBI and Rhea Endpoints:** Utilize the ChEBI and Rhea SPARQL endpoints to fetch the required information.\\n\\n**Example Endpoints:**\\n- ChEBI: https://www.ebi.ac.uk/rdf/services/sparql\\n- Rhea: https://sparql.rhea-db.org/sparql\\n\\n\\n### Generated SPARQL Query\\n\\n```sparql\\n# https://www.ebi.ac.uk/rdf/services/sparql\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX chebi: <http://purl.obolibrary.org/obo/CHEBI_>\\n\\nSELECT DISTINCT ?parent_reaction ?enzyme ?metabolite\\nWHERE {\\n  # Define the given metabolite\\n  VALUES ?metabolite {obo:CHEBI_83137}\\n  \\n  # Query Rhea to find reactions involving the given metabolite\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:side ?side .\\n    ?side ?p ?metabolite .\\n    ?reaction rhea:equation ?equation .\\n    ?reaction rhea:direction ?direction .\\n    \\n    # Identify the parent reaction\\n    ?parent_reaction rhea:reaction ?reaction .\\n  }\\n  \\n  # Query Rhea to find enzymes catalyzing the parent reaction\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?parent_reaction rhea:catalyzedBy ?enzyme .\\n  }\\n  \\n  # Optionally, fetch additional information about the metabolite from ChEBI\\n  SERVICE <https://www.ebi.ac.uk/rdf/services/sparql> {\\n    ?metabolite obo:IAO_0000115 ?description .\\n  }\\n}\\n```\\n\\nThis SPARQL query is designed to retrieve parent reactions and their catalyzing enzymes for a given metabolite (CHEBI:83137). It utilizes the Rhea SPARQL endpoint to identify reactions involving the metabolite and their parent reactions, as well as the enzymes catalyzing these parent reactions. The ChEBI endpoint is used to optionally fetch additional information about the metabolite. \\n\\nTo adapt this query for other metabolites, simply replace `obo:CHEBI_83137` with the appropriate ChEBI identifier for the metabolite of interest. \\n\\nThe query is federated, using the `SERVICE` keyword to distribute the query across the Rhea and ChEBI SPARQL endpoints, thus efficiently gathering the required information from these distinct databases. \\n\\nThe primary endpoint for initiating this federated query is the ChEBI SPARQL endpoint, as indicated by the comment on the first line of the query. \\n\\nPlease ensure that the URI for the given metabolite is correctly formatted and exists in the ChEBI database to avoid any errors. \\n\\nThis query adheres to the requirements by directly answering the user's question, providing comments to explain the logic, and utilizing the specified output format and endpoints.\", substeps=[], type='context', fixed_message=None)]}\n",
      "Got first run!\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Result: {'final_state_response': '# https://www.ebi.ac.uk/rdf/services/sparql\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX rhea: <http://rdf.rhea-db.org/>\\nPREFIX chebi: <http://purl.obolibrary.org/obo/CHEBI_>\\n\\nSELECT DISTINCT ?parent_reaction ?enzyme ?metabolite\\nWHERE {\\n  # Define the given metabolite\\n  VALUES ?metabolite {obo:CHEBI_83137}\\n  \\n  # Query Rhea to find reactions involving the given metabolite\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rhea:side ?side .\\n    ?side ?p ?metabolite .\\n    ?reaction rhea:equation ?equation .\\n    ?reaction rhea:direction ?direction .\\n    \\n    # Identify the parent reaction\\n    ?parent_reaction rhea:reaction ?reaction .\\n  }\\n  \\n  # Query Rhea to find enzymes catalyzing the parent reaction\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?parent_reaction rhea:catalyzedBy ?enzyme .\\n  }\\n  \\n  # Optionally, fetch additional information about the metabolite from ChEBI\\n  SERVICE <https://www.ebi.ac.uk/rdf/services/sparql> {\\n    ?metabolite obo:IAO_0000115 ?description .\\n  }\\n}', 'run_id_langsmith': '7405426c-76a2-44b1-8a6e-65e5d1770bb8', 'in_dataset': False, 'execution_time': '0', 'sparql_construction_prompt_tokens': 705, 'sparql_construction_completion_tokens': 0, 'sparql_construction_total_tokens': 705, 'sparql_construction_prompt_cost': 0, 'sparql_construction_completion_cost': 0, 'sparql_construction_total_cost': 0, 'prompt_tokens': 1259, 'completion_tokens': 86, 'total_tokens': 1345, 'prompt_cost': 0, 'completion_cost': 0, 'total_cost': 0}\n",
      "Got result\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/5_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/50_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/51_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/12_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/112_Total_number_of_reaction_ec_protein_links_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/eval/ev_2025-04-08_19-56-53/117_Retrieve_parent_reactions_and_their_catalyzing_enzymes_for_a_metabolite_that_is_not_directly_involved_in_a_reaction_comparison.ttl\n"
     ]
    }
   ],
   "source": [
    "\n",
    "evaluator = AgentEvaluator(\n",
    "    endpoint_sets=test_set,\n",
    "    experiment_dir=\"experiments_official\",\n",
    "    project_name_langsmith=\"sparql-rag-agent-baseline-version-1.0\"\n",
    ")\n",
    "evaluation_dataset_path = await evaluator.run_all_tests()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
